### PR TITLE
Fix Numberfield android focus lost

### PR DESCRIPTION
**Pull request by snowystinger**
Closes https://github.com/adobe/react-spectrum/issues/2119

This may bring up the software keyboard after the press event ends. This is probably ok because we were trying to avoid it opening on start which would shift the button out from under the user's finger, canceling the spinning behavior.
Another option might be to give focus to the group, but this might be weird because technically the group isn't focusable and we don't have a ref to it.

## ✅ Pull Request Checklist:

- [ ] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [ ] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [ ] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/TR/wai-aria-practices-1.1/)

### CODE REVIEW

**Review by LFDanLu — APPROVED** [review]
LGTM, tested on my android and on an iPad. I think the behavior is ok.

I ran into a weird issue on the build storybook where pressing the increment/decrement buttons via Voiceover double tap was causing focus to shift away from the button to the input but I couldn't reproduce it locally or after a refresh so might be something transient. Not a blocker unless others can reproduce

**Comment by Doesntmeananything** [comment]
I'm seeing something strange in the storybook build on my android device when holding down a stepper button with the virtual keyboard open: after holding the button for about a second the keyboard closes by itself and then reopens again. I think holding down the tap triggers some additional event that interferes here. But I can't say for sure since storybook is a bit laggy on my device.

@LFDanLu these are the steps to reproduce:

1. Go to a NumberField example, e.g. here: https://reactspectrum.blob.core.windows.net/reactspectrum/b0fb4c74891469d57c4cf0c035e3f7fee2166c46/storybook/index.html?path=/story/numberfield--minvalue-0-maxvalue-20
2. Tap on the input field to open the virtual keyboard.
3. Tap and hold any stepper buttons.


https://user-images.githubusercontent.com/34657799/126466371-9791b006-c375-4a29-afe1-99d19e0a3d07.mp4



I would expect the keyboard to stay open, which is what I'm currently seeing in the offical docs and locally with the latest react-aria release.

I also think the change in behaviour to focus the input field on press end is too drastic. Virtual keyboard can feel very disruptive, and opening after taps on stepper buttons, to me, is counter intuitive: 
- A user may have finished interacting with the NumberField after one tap, yet they're now presented with a keyboard to edit the field value.
- Doing multiple taps is kind of janky - if a user wants to increment field value by 3, and they do 3 consecutive taps, chances are the keyboard will appear and push the button away.

I think you did a fantastic job implementing the previous behavior (no keyboard unless the input field is explicitly focused) and I'm hoping that it can stay.

**Comment by snowystinger** [comment]
Thanks @Doesntmeananything, this is valuable feedback

**Comment by LFDanLu** [comment]
@Doesntmeananything Good catch with virtual keyboard closing while holding the stepper button (only seems to happen in the storybook iframe). Perhaps there is a way we can make it so the virtual keyboard doesn't show up on stepper button tap, think I had something in ComboBox that did something similar in a previous iteration.

**Comment by snowystinger** [comment]
This PR has tried to address a few issues:
1. Focus ring sticks around on Android chrome NumberField after pressing and reaching the max/min and the button disables under the users finger.
2. Focus on a disabled button is lost to body
3. The software keyboard shouldn't interfere with a press and hold
4. The software keyboard shows up after the user lets go of the increment/decrement button on a long hold

In the case of #1, the focus ring sticks around because a disabled button won't fire blur events in chrome.
We've advised people in the past to make sure to focus an appropriate element before the disable is applied.

The other 3 require a bit of a story.
When in a situation where a software keyboard may be present, it should not interfere with the increment/decrement buttons. What this means is that if the keyboard is already up, it should stay up, and if it's down, it should stay down for as long as the user is interacting with those buttons.
We accomplish this by putting focus on the button being held, not on the input like we do in all other cases.
The problem arises when the user has long pressed and the button they were holding down on becomes disabled. At this point, the focus is lost to the document.body. We can't send focus to the next or previous inputs, we only know about the NumberField. So we send focus to the input.
This has a side effect of opening the keyboard in most cases because we move the focus in response to a user driven event.

We tried numerous ways to move the focus from the button to the input at a better time so that #1 wouldn't be an issue we needed to solve. However, our attempts resulted in unsafe flushes during render.
- Tried inline in useNumberField
```
  let hasDisabled = !state.canIncrement || !state.canDecrement;
  let hadDisabled = useRef(hasDisabled);

  if (document.activeElement !== inputRef.current && focusWithin && !hadDisabled.current && hasDisabled) {
    inputRef.current.focus(); // causes a state change in the middle of a render
  }
  hadDisabled.current = hasDisabledt;
```

- Tried it inside useNumberFieldState with a listener approach
```
// useNumberFieldState
  let listeners = useRef([]);
  let aboutToDisableButton = (fun) => {
    listeners.current.push(fun);
  }
...
  let increment = () => {
      ...
      if (newValue === maxValue) {
        listeners.current.forEach(fun => fun());
      }

// useNumberField
useEffect(() => {
  state.register(() => inputRef.current.focus()); // still resulted in state change during render?
}, []);
```

It's possible we may be able to eventually use navigator.virtualKeyboard, but that spec is likely a long ways off.

We also tried this, which keeps the keyboard down when tapping or long pressing, but breaks if the keyboard is already open.
```
  let onButtonPressEnd = (e) => {
    inputRef.current.readOnly = true;
    if (e.pointerType !== 'mouse') {
      inputRef.current.focus();
    }
    setTimeout(() => {
      inputRef.current.readOnly = false;
    })
  };
```

**Review by devongovett — APPROVED** [review]

File: packages/@react-aria/button/src/useButton.ts [hunk 1]
```diff
@@ -54,6 +54,8 @@ export function useButton(props: AriaButtonProps<ElementType>, ref: RefObject<an
     onPressChange,
     // @ts-ignore - undocumented
     preventFocusOnPress,
+    // @ts-ignore - undocumented
+    allowFocusWhenDisabled,
```
Comment (devongovett) [inline packages/@react-aria/button/src/useButton.ts:58]: 😬

Comment (devongovett) [reply packages/@react-aria/button/src/useButton.ts:58]: Let's hope no one abuses this to put tooltips on disabled things...

Comment (snowystinger) [reply packages/@react-aria/button/src/useButton.ts:58]: why would you call it out then? =P we can always just change the name of the prop every couple releases to really discourage it, hahaha

**Pull request merged**